Make the Docker build more re-usable in Cloud

pugnascotia opened the pull request:
Closes #49926. Closes #46166.

   * Add a mini-init process
   * Don't use root at all when running the container
   * Ensure no files in the image have the setuid flag

Also improve dependency tracking in the build.

pugnascotia committed:
Make the Docker build more re-usable in Cloud

Closes #49926 and #46166.

   * Add a mini-init process (copied from Cloud)
   * Don't use root at all when running the container
   * Add an explicity TERM handler
   * Ensure no files in the image have the setuid flag

Also improve dependency tracking in the build.

pugnascotia commented:
cc @mieciu

jordansissel commented on distribution/docker/src/docker/bin/my_init:
```diff
@@ -0,0 +1,424 @@
+#!/usr/bin/python2 -u
```
Given this is a WIP; I have a question about the design here.

Can Elasticsearch be taught to handle SIGCHLD instead of adding a new process? Or are there other benefits to having this my_init thing? It's a lot of code with a bunch of features (ssh keys, sysv-style init scripts, etc) and maybe it can be simplified?

pugnascotia replied:
For now,tThis is literally a copy of what Cloud uses, but I'd like to replace it e.g. either by making ES more robust, or using something super-focussed like [tini](https://github.com/krallin/tini).

jordansissel commented on distribution/docker/src/docker/bin/docker-entrypoint.sh:
```diff
@@ -57,17 +41,35 @@ if [[ -f bin/elasticsearch-users ]]; then
+
+# Use a mini-init process to ensure any children are cleaned up and we
+# aren't left with any zombies.
+/sbin/my_init --skip-runit -- /usr/share/elasticsearch/bin/elasticsearch "$@" -p $PID_FILE &
```
these last few lines (background launch my_init, get pid, wait, exit w/ code) can be replaced by using exec.

```
exec /sbin/my_init --skip-runit -- /usr/share/elasticsearch/bin/elasticsearch "$@" -p $PID_FILE
```

For SIGTERM handling; moving that into `my_init` would help as well. No need to capture the PID of Elasticsearch either since `my_init` already knows what it is because it forked a process for it.

pugnascotia committed:
Use tini as the init system

Remove the my_init script in favour of tini, which meant that
docker-entrypoint.sh could be cleaned up significantly.

pugnascotia commented:
@jordansissel I've replaced the `my_init` script with [init](https://github.com/krallin/tini), and removed a lot of the dancing around that the entrypoint was doing.

Regarding SIGCHLD, I believe I'm correct in saying that ES treats unexpected termination of ML nodes as a problem, and doesn't restart them. IOW, ES doesn't expect to have to do any child reaping in the normal course of things. It might be worth a conversation with the ML team, my gut feeling is that it's perhaps orthogonal to this change?

I still need to make docs changes if we're broadly OK with these changes.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Packaging)

pugnascotia committed:
Add test for PID 1 in Docker container

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 49926-docker-changes-for-ess

pugnascotia commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into 49926-docker-changes-for-ess

pugnascotia committed:
Also remove setgid bit in Dockerfile

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 49926-docker-changes-for-ess

dliappis commented on distribution/docker/src/docker/bin/docker-entrypoint.sh:
```diff
@@ -56,18 +40,13 @@ if [[ -f bin/elasticsearch-users ]]; then
 
-if [[ "$(id -u)" == "0" ]]; then
-  # If requested and running as root, mutate the ownership of bind-mounts
-  if [[ -n "$TAKE_FILE_OWNERSHIP" ]]; then
```
We'll need to adjust the [docs](https://github.com/elastic/elasticsearch/blob/master/docs/reference/setup/install/docker.asciidoc) and remove the `TAKE_FILE_OWNERSHIP` description as mentioned in the 3rd task in https://github.com/elastic/elasticsearch/issues/46166.

pugnascotia replied:
Ah, yes, I meant to remove that from the docs. Good spot 👍

dliappis reviewed (changes requested):
Thank you!

This is wonderful work and cleans things up so much.
I asked for a change because we need to update the docs to reflect the removal of TAKE_FILE_OWNERSHIP.

Additionally I think @droberts195 needs to review this to verify that the order of killing is correct (as discussed in https://github.com/elastic/elasticsearch/issues/49926#issuecomment-565512109)

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 49926-docker-changes-for-ess

pugnascotia committed:
Remove TAKE_FILE_OWNERSHIP from documentation

droberts195 reviewed (approved):
From https://github.com/krallin/tini#process-group-killing I see that:

> By default, Tini only kills its immediate child process.

So LGTM from the ML perspective.

pugnascotia commented:
Aye, I went digging for that exact information before 👍

dliappis reviewed (approved):
LGTM.

I mulled over whether we should mark the removal of `TAKE_FILE_OWNERSHIP` as a breaking change (`reference/migration/migrate_8_0.asciidoc`) but it really was meant to be used as a last resort as it mutated the bind mount permissions from within the container, so possibly not a big deal.

pugnascotia committed:
Ensure functional permissions when bind-mounting for DockerTests

pugnascotia commented:
@dliappis I had to adjust the tests so that the file permissions when bind-mounting directories were acceptable. In the process, I added more checking into the `elasticsearch-env-from-file` script to diagnose such problems, since the message e.g. "File /run/secrets/esJavaOpts.txt from ES_JAVA_OPTS_FILE does not exist" is misleading.

dliappis commented on distribution/src/bin/elasticsearch-env-from-file:
```diff
@@ -20,19 +20,43 @@ for VAR_NAME_FILE in $(env | cut -f1 -d= | grep '_FILE$'); do
+      # Maybe the file doesn't exist, maybe we just can't read it due to file permissions.
+      # Check permissions on each part of the path
+      if ! echo "${!VAR_NAME_FILE}" | grep -q '^/'; then
+        path=.
```
If this conditional doesn't execute what is the value of `path`? AFAICS we rely on some `$path` value in the following `while` loop (lines 29-34)

pugnascotia replied:
The value is empty. I've added an explicit initialisation for this.

dliappis reviewed (commented):
I left two questions.

pugnascotia committed:
Fix DockerTests when not running as root

Also explicitly initialise a variable in elasticsearch-env-from-file

pugnascotia commented on qa/os/src/test/java/org/elasticsearch/packaging/util/Docker.java:
```diff
@@ -152,9 +153,17 @@ private static void executeDockerRun(Distribution distribution, Map<Path, Path>
+
+                // The process in the Docker container doesn't run as root, so we need to ensure that
+                // it is able to read the bind-mounted files.
+                sh.run("chown -R 1000:0 " + localPath);
```
It does indeed. Our tests in CI run as root, which is why I had a problem with permissions in the first place, and why this changes fixes CI, but it also breaks the tests when running locally as non-root. I've wrapped this with an additional check to fix the problem.

dliappis reviewed (approved):
LGTM thanks

pugnascotia committed:
Merge remote-tracking branch 'upstream/master' into 49926-docker-changes-for-ess

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix-unix due to transient error during sles build.

pugnascotia committed:
Check tini with sha512 instead of gpg

tini's webpack suggests validating the binary with gpg, but the
keyservers appear to be slow and unreliable. Swap this check with
a sha512 check.

pugnascotia commented:
@elasticmachine run elasticsearch-ci/packaging-matrix-unix due to sles transient failures
